### PR TITLE
Updates + optimized moshi + kryo

### PR DESCRIPTION
**Pull request by ZacSweers**
Updated dependencies, some cleanup

- Added optimized moshi example (where adapter cache is already hot)
- Add kryo sample
  - I can't get this to work. Kryo's API makes no sense to me and it complains of not finding a serializer despite me registering all the classes. @naturalwarren @erichkleung any suggestions here?

### CODE REVIEW

**Comment by naturalwarren** [comment]
Pulling this in to take a closer peak locally. I'm excited for this PR.

File: src/jmh/java/com/example/SpeedTest.java [hunk 1]
```diff
@@ -78,10 +91,31 @@ public void doSetup() throws Exception {
         public ResponseAV response;
     }
 
+    // An optimized moshi instance where the cache is already hot
```
Comment (naturalwarren) [inline src/jmh/java/com/example/SpeedTest.java]: Would we be interested in a comparison with a warm Gson cache?

File: src/jmh/java/com/example/SpeedTest.java [hunk 2]
```diff
@@ -78,10 +91,31 @@ public void doSetup() throws Exception {
...
+                    .add(GeneratedJsonAdapterFactory.create())
+                    .build();
+            URL url = Resources.getResource("largesample.json");
+            json = Resources.toString(url, Charsets.UTF_8);
+            response = moshi
+                    .adapter(ResponseAV.class)
+                    .fromJson(json);
+        }
+
+        public Moshi moshi;
```
Comment (naturalwarren) [inline src/jmh/java/com/example/SpeedTest.java]: Out of curiosity why did we define these as member variables below the setup method?

File: build.gradle [hunk 3]
```diff
@@ -28,18 +28,24 @@ repositories {
...
-    compileOnly 'com.google.auto.value:auto-value:1.4-rc2'
+    compileOnly 'com.google.auto.value:auto-value:1.4.1'
     compileOnly 'com.ryanharter.auto.value:auto-value-gson:0.4.6'
     compileOnly 'com.ryanharter.auto.value:auto-value-moshi:0.4.3'
 
-    jmh 'com.google.guava:guava:21.0'
-    compile 'com.squareup.okio:okio:1.11.0'
-    compile 'com.squareup.moshi:moshi:1.4.0'
-    compile 'com.google.code.gson:gson:2.8.0'
+    jmh 'com.google.guava:guava:22.0'
```
Comment (naturalwarren) [inline build.gradle:39]: Neat!

**Comment by naturalwarren** [comment]
For Kryo how about we do something like the following:

```java
  public static void main(String[] args) throws IOException, ClassNotFoundException {
        Kryo kryo = new Kryo();
        Kryo.DefaultInstantiatorStrategy instantiatorStrategy = new Kryo.DefaultInstantiatorStrategy();
        instantiatorStrategy.setFallbackInstantiatorStrategy(new StdInstantiatorStrategy());
        kryo.setInstantiatorStrategy(instantiatorStrategy);

        URL url = Resources.getResource("largesample.json");
        String json = Resources.toString(url, Charsets.UTF_8);
        ResponseAV response = new GsonBuilder()
                .registerTypeAdapterFactory(GeneratedTypeAdapterFactory.create())
                .create()
                .fromJson(json, ResponseAV.class);

        ByteArrayOutputStream stream = new ByteArrayOutputStream();
        Output output = new Output(stream);
        kryo.writeObject(output, response);
        output.flush();

        byte[] bytes = stream.toByteArray();
        ResponseAV o = (ResponseAV) kryo.readObject(new Input(bytes),
                getAutoValueClass(ResponseAV.class, ResponseAV.class.getPackage()));
    }

    public static Class getAutoValueClass(Class<?> clazz, Package pkg) throws ClassNotFoundException {
        String pkgName = pkg == null ? "" : pkg.getName();
        String className = pkgName + ".AutoValue_" + clazz.getSimpleName();
        return Class.forName(className);
    }
```

File: src/jmh/java/com/example/SpeedTest.java (hunk 1 above)
Comment (ZacSweers) [reply src/jmh/java/com/example/SpeedTest.java]: I'm not particularly interested

File: src/jmh/java/com/example/SpeedTest.java (hunk 2 above)
Comment (ZacSweers) [reply src/jmh/java/com/example/SpeedTest.java]: Just how JMH works. Stuff done in `@Setup` won't be part of the benchmark

**Comment by ZacSweers** [comment]
that example doesn't work for me, I just get this during setup when trying to call `writeObject`

```
java.lang.IllegalArgumentException: serializer cannot be null.
        at com.esotericsoftware.kryo.Registration.<init>(Registration.java:37)
        at com.esotericsoftware.kryo.util.DefaultClassResolver.registerImplicit(DefaultClassResolver.java:74)
        at com.esotericsoftware.kryo.Kryo.getRegistration(Kryo.java:508)
        at com.esotericsoftware.kryo.Kryo.writeObject(Kryo.java:557)
        at com.example.SpeedTest.kryo_toBytes(SpeedTest.java:229)
```

**Comment by naturalwarren** [comment]
> that example doesn't work for me

I'm not sure then. That sample runs for me and results in successful deserialization of `ResponseAV`. Can you double check that your setup reflects exactly what was pasted?

**Comment by naturalwarren** [comment]
Preliminary results from my machine below. I can run the full suite at some point today or provide the Kryo benchmark code if you like. 

```
Result "kryo_fromBytes":
  2809.756 ±(99.9%) 32.564 ops/s [Average]
  (min, avg, max) = (2188.404, 2809.756, 3068.536), stdev = 137.880
  CI (99.9%): [2777.192, 2842.321] (assumes normal distribution)


# Run complete. Total time: 00:06:56

Benchmark                  Mode  Cnt     Score    Error  Units
SpeedTest.kryo_fromBytes  thrpt  200  2809.756 ± 32.564  ops/s
```

**Comment by ZacSweers** [comment]
Mind pasting a diff of your local changes?

**Comment by erichkleung** [comment]
@naturalwarren is that a reflective kryo fromBytes? Or with a pre-written serializer? (Looks like non-reflective, non-streaming) 

If I looked at it right, that was 1400 for moshi streaming vs 2800 with kryo.

**Comment by ZacSweers** [comment]
also different machines and conditions. I also want to make sure all the fields are deserializing correctly, as we had false positives before where it was just failing early on fields and skipping them all

**Comment by naturalwarren** [comment]
Git patch: 
[Patch.txt](https://github.com/hzsweers/json-serialization-benchmarking/files/1073158/Patch.txt)

**Comment by ZacSweers** [comment]
Ok some interesting updates:

```
Benchmark                                      Mode  Cnt     Score   Error  Units
SpeedTest.gson_reflective_fromJson            thrpt  200  1413.360 ± 6.983  ops/s
SpeedTest.gson_reflective_toJson              thrpt  200   981.531 ± 7.825  ops/s
SpeedTest.gson_streaming_fromJson             thrpt  200  1330.630 ± 6.014  ops/s
SpeedTest.gson_streaming_toJson               thrpt  200  1163.326 ± 6.307  ops/s
SpeedTest.kryo_fromBytes                      thrpt  200  1314.428 ± 4.732  ops/s
SpeedTest.kryo_toBytes                        thrpt  200   933.753 ± 3.615  ops/s
SpeedTest.moshi_reflective_fromJson           thrpt  200   840.123 ± 4.572  ops/s
SpeedTest.moshi_reflective_toJson             thrpt  200  1330.161 ± 7.834  ops/s
SpeedTest.moshi_streaming_fromJson            thrpt  200   899.152 ± 6.889  ops/s
SpeedTest.moshi_streaming_optimized_fromJson  thrpt  200   887.064 ± 9.336  ops/s
SpeedTest.moshi_streaming_optimized_toJson    thrpt  200  1437.873 ± 6.703  ops/s
SpeedTest.moshi_streaming_toJson              thrpt  200  1436.459 ± 5.250  ops/s
```

- Something is wrong with those gson results. Going to investigate more, but seeing as there was nothing substantial from 2.8.0 -> 2.8.1 I'm thinking something's weird here. I suspect something is error-ing out during writes, leaving us with missing data and subsequently even faster json reads. Should be easy to check.
- Kryo is substantially slower with `CompatibleFieldSerializer`. Its writes in particular are super slow, but its reads are still faster relative to the others. 46% faster than streaming moshi.
- "Moshi optimized" is not really optimized. Heating up the adapter cache first didn't make much of a difference, but this isn't quite reflect of the state we're usually in either (explained below).

Further questions:
- It would be interesting to see how Kryo reads are affected by field changes
- I think a more accurate test for moshi would be to push the json string into a buffer first before starting the benchmark, since in Retrofit we're already reading from the Response's `BufferedSource`. I tried doing this at first but it seems that some internal state gets lost in the jmh tests and I'm not reaaaaaaally sure what to do about it. It just EOF's early if I create the buffer in the setup method rather than wait till the benchmark (even if I `clone()` it first). I might be able to measure that cost independently though, and thus add it back to the optimized fromJson() measurement above.

**Comment by ZacSweers** [comment]
> If I looked at it right, that was 1400 for moshi streaming vs 2800 with kryo.

@erichkleung That was not using CompatibleFieldSerializer btw. Makes sense it's twice as fast since it's doing half the work (not writing any form of key or having backward compatibility)

**Comment by ZacSweers** [comment]
Update on the buffer testing: I was able to measure the buffer cost separately and calculate it as a percentage of the main:

```
Benchmark                                     Mode  Cnt     Score    Error  Units
SpeedTest.moshi_streaming_optimized_buffer    avgt  200   140.587 ±  8.531  us/op
SpeedTest.moshi_streaming_optimized_fromJson  avgt  200  1141.428 ± 10.338  us/op
```

Which gives us `12.3167646%`. If we apply this to the `fromJson` above, we can extrapolate 

`moshi_streaming_fromJson` -> `1009.898435236`
`moshi_streaming_optimized_fromJson` -> `996.321584731`

So if we update the earlier results:

```
SpeedTest.gson_reflective_fromJson            thrpt  200  1413.360 ± 6.983  ops/s
SpeedTest.gson_reflective_toJson              thrpt  200   981.531 ± 7.825  ops/s
SpeedTest.gson_streaming_fromJson             thrpt  200  1330.630 ± 6.014  ops/s
SpeedTest.gson_streaming_toJson               thrpt  200  1163.326 ± 6.307  ops/s
SpeedTest.kryo_fromBytes                      thrpt  200  1314.428 ± 4.732  ops/s
SpeedTest.kryo_toBytes                        thrpt  200   933.753 ± 3.615  ops/s
SpeedTest.moshi_reflective_fromJson           thrpt  200   840.123 ± 4.572  ops/s
SpeedTest.moshi_reflective_toJson             thrpt  200  1330.161 ± 7.834  ops/s
SpeedTest.moshi_streaming_fromJson            thrpt  200  1009.898  ops/s
SpeedTest.moshi_streaming_optimized_fromJson  thrpt  200   996.322  ops/s
SpeedTest.moshi_streaming_optimized_toJson    thrpt  200  1437.873 ± 6.703  ops/s
SpeedTest.moshi_streaming_toJson              thrpt  200  1436.459 ± 5.250  ops/s
```

**Comment by ZacSweers** [comment]
Some final numbers

```
Benchmark                                      Mode  Cnt     Score    Error  Units
SpeedTest.gson_reflective_fromJson            thrpt  200  1387.786 ± 11.667  ops/s
SpeedTest.gson_reflective_toJson              thrpt  200   954.493 ± 11.019  ops/s
SpeedTest.gson_streaming_fromJson             thrpt  200  1299.684 ± 12.719  ops/s
SpeedTest.gson_streaming_toJson               thrpt  200  1162.031 ±  8.967  ops/s
SpeedTest.kryo_fromBytes                      thrpt  200  1175.603 ±  5.007  ops/s
SpeedTest.kryo_toBytes                        thrpt  200   895.594 ±  4.118  ops/s
SpeedTest.moshi_reflective_fromJson           thrpt  200   810.387 ±  6.878  ops/s
SpeedTest.moshi_reflective_toJson             thrpt  200  1346.315 ±  4.937  ops/s
SpeedTest.moshi_streaming_fromJson            thrpt  200   858.326 ± 11.060  ops/s
SpeedTest.moshi_streaming_fromJson_optimized  thrpt  200  1012.564 ±  5.320  ops/s
SpeedTest.moshi_streaming_toJson              thrpt  200  1457.232 ±  9.886  ops/s
SpeedTest.moshi_streaming_toJson_optimized    thrpt  200  1687.757 ± 13.782  ops/s
```

**Comment by ZacSweers** [comment]
so the gist of it 

- no idea what's up with gson. Those numbers don't make sense but I can't find anything to indicate that they're wrong.
- buffered setup for moshi "optimized" reflects more of the average stack since it's just reading in from or writing to buffers.

**Comment by ZacSweers** [comment]
the json reading improvements in gson seem to be from https://github.com/google/gson/commit/410b4a8c62de693d38a0ee9a566a4a4c0863f934

**Comment by naturalwarren** [comment]
> I think a more accurate test for moshi would be to push the json string into a buffer first before starting the benchmark, since in Retrofit we're already reading from the Response's BufferedSource. I tried doing this at first but it seems that some internal state gets lost in the jmh tests and I'm not reaaaaaaally sure what to do about it. It just EOF's early if I create the buffer in the setup method rather than wait till the benchmark (even if I clone() it first). I might be able to measure that cost independently though, and thus add it back to the optimized fromJson() measurement above.

This may reflect our setup for network (de)serialization but certainly not disk - we'd be handed a POJO. 

A few thoughts on the speeds here within the context of switching from Kryo to Moshi/Gson for disk (de)serialization:
- We probably don't care about write speed. When writing something to disk we usually don't, if ever, care about how long a background thread blocks for
- Moshi read is 31% poorer than Kryo read (858 vs 1175)
- Gson read is 10% better than Kryo read (1299 vs 1175)

For completion's sake it would be nice to have streaming Kryo benchmarks. @hzsweers Perhaps I could contribute these?

**Comment by ZacSweers** [comment]
Added gson buffer streaming

```
Benchmark                                   Mode  Cnt     Score    Error  Units
SpeedTest.gson_reflective_fromJson         thrpt  200  1338.221 ± 14.028  ops/s
SpeedTest.gson_reflective_toJson           thrpt  200   960.946 ±  4.541  ops/s
SpeedTest.gson_streaming_fromJson          thrpt  200  1309.312 ± 11.673  ops/s
SpeedTest.gson_streaming_fromJson_buffer   thrpt  200  1174.803 ±  4.228  ops/s
SpeedTest.gson_streaming_toJson            thrpt  200  1129.538 ±  4.990  ops/s
SpeedTest.gson_streaming_toJson_buffer     thrpt  200   593.499 ±  2.120  ops/s
SpeedTest.kryo_fromBytes                   thrpt  200  1286.459 ±  5.362  ops/s
SpeedTest.kryo_toBytes                     thrpt  200   921.524 ±  5.121  ops/s
SpeedTest.moshi_reflective_fromJson        thrpt  200   843.788 ±  7.334  ops/s
SpeedTest.moshi_reflective_toJson          thrpt  200  1375.238 ±  6.176  ops/s
SpeedTest.moshi_streaming_fromJson         thrpt  200   869.343 ±  6.845  ops/s
SpeedTest.moshi_streaming_fromJson_buffer  thrpt  200   980.322 ±  5.890  ops/s
SpeedTest.moshi_streaming_toJson           thrpt  200  1421.798 ±  6.761  ops/s
SpeedTest.moshi_streaming_toJson_buffer    thrpt  200  1666.212 ± 13.413  ops/s
```

**Pull request merged**